runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/internal/registry/NamespaceRegistry.java

```diff
@@ -153,13 +154,28 @@ private void setMetricHandlers(
                 LongConsumer metricRecorder = supplyMetricRecorder.apply(metric.kind(), config.id, metricId);
                 MessageConsumer handler = metric.supplyHandler(metricRecorder);
                 MetricHandlerKind kind = resolveKind(binding.originTypeId(), binding.routedTypeId(), metric.group());
+                Metric.StreamDirection direction = metric.streamDirection();
                 if (kind == ROUTED)
                 {
-                    routedMetricHandler = routedMetricHandler.andThen(handler);
+                    if (direction == SENT || direction == BOTH)
+                    {
+                        sentRoutedMetricHandler = sentRoutedMetricHandler.andThen(handler);
+                    }
+                    if (direction == RECEIVED || direction == BOTH)
+                    {
+                        receivedRoutedMetricHandler = receivedRoutedMetricHandler.andThen(handler);
+                    }
                 }
                 else if (kind == ORIGIN)
                 {
-                    originMetricHandler = originMetricHandler.andThen(handler);
+                    if (direction == SENT || direction == BOTH)
+                    {
+                        sentOriginMetricHandler = sentOriginMetricHandler.andThen(handler);
+                    }
+                    if (direction == RECEIVED || direction == BOTH)
+                    {
+                        receivedOriginMetricHandler = receivedOriginMetricHandler.andThen(handler);
+                    }
```

Comment:
I feel like there should be a simpler way to handle this than nested `if-else`, but it works. 🙂

Reply:
Yeah, unfortunately it looks like this, but I don't exactly see a better way. We could introduce a map, but we have two keys so it should be something like `Map<Direction, Map<MetricHandlerKind, MessageConsumer>>` so I am not sure it would be actually simpler. 🤔